File: py/Recursion.py
cache = {}


def memo(fn, arg: int) -> int:
    if arg not in cache:
        cache[arg] = fn(arg)
    return cache[arg]


def fibonaci(n: int) -> int:
    if n == 0:
        return 0
    if n == 1:
        return 1
    else:
        return memo(fibonaci, n-1) + memo(fibonaci, n-2)


def factorial(x: int) -> int:
    if x == 0:
        return 1
    else:
        return x * memo(factorial, x-1)

File: rust/Recursion.rs
use std::collections::HashMap;

fn main() {
    let cache: HashMap<u64, u64> = HashMap::new();

    
   fn fibonnaci(n: u64) -> u64 {
        if n <= 1 {
            return n;
        }

        return fibonnaci(n - 1) + fibonnaci(n - 2);
    }

    fn factorial(n: u64) -> u64 {
        if n <= 0 {
            return 1;
        }

        return n * factorial(n - 1);
    }
}
